tst/com/amazon/corretto/crypto/provider/test/EvpSignatureSpecificTest.java

```diff
@@ -638,6 +638,36 @@ public void ecdsaSignCorruptsErrorState() throws Exception {
     }
   }
 
+  /**
+   * For better compatibility with the JDK, we want to throw SignatureException if the length of the
+   * signature is wrong.
```

Comment:
this may help with JDK compatibility, but it diverges from [BC's behavior](https://github.com/bcgit/bc-java/blob/91c3c6018ab353f653749b9b56d6de384a31ec3c/prov/src/main/java/org/bouncycastle/jcajce/provider/asymmetric/rsa/DigestSignatureSpi.java#L146) of returning `false` rather than throwing.

hoping to have further discussion here with greg and the ACCP team on tradeoffs of favoring JDK compatibility vs. BC compatibility.

Reply:
In an ideal world, it would only be `false` or `true`, but unfortunately that isn't the world we're in.

We don't need to worry about "leaking information" because every piece of data handled here is public. (Signatures are considered public and public keys are used to validate them.) The concerns are complexity and compatibility.

Complexity: Developers already (unfortunately) need to handle `SignatureException` and this doesn't make things any worse for them.

Compatibility: The JDK's providers (and tests) define the standard through their behaviors. ACCP cannot match it *perfectsly*, but it can try to minimize the differences. Given a choice between mutually exclusive options of "compatible with JDK providers" and "compatible with BC," I believe that ACCP should *usually* choose to be compatible with the JDK providers. (Obviously, there may be questions of security and correctness, but neither apply here.)

Finally, I believe it is a reasonably generalization to say "Teams use the JDK providers by default and might *choose* to use BC." This means that use of BC is *more likely* to be an intentional decision rather than using the JDK providers. So, if ACCP is being added, we shouldn't change how things work for teams which are just grabbing the default and not paying attention to the details. If a team has chosen to use BC then:

* They are overriding ACCP (so these changes don't matter anyway), OR
* They are making intentional decisions around their cryptography and so can be expected to make adjustments to their code if needed.

Truthfully, many teams use BC without thinking about it (either because of a dependency or just not understanding cryptographic code). But it remains true that using BC is more of a conscious choice than just sticking with the default.

Reply:
ACCP-FIPS compatibility with BCFIPS is highly desirable. Customers that have FIPS requirements and want to switch to ACCP-FIPS for performance and still satisfy the FIPS requirements would have simpler transition.
P.S. we have just released ACCP-FIPS for experimental purposes. AWS-LC-FIPS is still waiting for it certificate.

Reply:
I still believe that my above positions hold the vast majority of the time.

Careful code *needs* to handle `SignatureException` (partially because it is a checked exception).

There is an interesting question as to whether there should be some way to tune ACCP's compatibility towards JDK or BC, but I think the answer there should remain "no."

Reply:
I agree with @SalusaSecondus's points. The incompatibility introduced here is trivially resolved by the customers and returning more contextual information in our SignatureExceptions is more helpful than a boolean failure.

Reply:
I agree with @SalusaSecondus's points. The incompatibility introduced here is trivially resolved by the customers and returning more contextual information in our SignatureExceptions is more helpful than a boolean failure.

Reply:
Sounds good to me.

Reply:
Will's guideline is a great mechanism for deciding how incompatibilities should be handled.